### PR TITLE
fix(NODE-5551): set AWS region from environment variable for STSClient

### PR DESCRIPTION
**Pull request by nbbeeken**
#### What is changing?

- Adding detection for `AWS_STS_REGIONAL_ENDPOINTS` and `AWS_REGION` environment variables and setting the corresponding `fromNodeProviderChain` options.
- `LEGACY_REGIONS` is created from this: https://docs.aws.amazon.com/sdkref/latest/guide/feature-sts-regionalized-endpoints.html documentation
- The driver will only set region options if BOTH environment variables are present, this is so that the driver does not encode a default setting for `AWS_STS_REGIONAL_ENDPOINTS`.

##### Is there new documentation needed for these changes?

Yes. Will file a docs ticket.

#### What is the motivation for this change?

AWS recommends setting your STS region to be as close to your app as possible to reduce latency but currently, the JS SDK does not use the common AWS region environment variables, and the setting must be controlled programmatically.

Tracking issue here: https://github.com/aws/aws-sdk-js-v3/issues/5105

### Use region settings for STS AWS credentials request

When using [IAM AssumeRoleWithWebIdentity](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRoleWithWebIdentity.html) AWS authentication the driver uses the [@aws-sdk/credential-providers](https://github.com/aws/aws-sdk-js-v3) package to contact the Security Token Service API for temporary credentials. AWS recommends using Regional AWS STS endpoints instead of the global endpoint to reduce latency, build-in redundancy, and increase session token validity. Unfortunately, environment variables `AWS_STS_REGIONAL_ENDPOINTS` and `AWS_REGION` do not directly control the region the SDK's STS client contacts for credentials. 

The driver now has added support for detecting these variables and setting the appropriate options when calling the SDK's API: [fromNodeProviderChain()](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/Package/-aws-sdk-credential-providers/#fromNodeProviderChain). 

> [!IMPORTANT] 
> The driver will only set region options if **BOTH** environment variables are present. `AWS_STS_REGIONAL_ENDPOINTS` **MUST** be set to either `'legacy'` or `'regional'`, and `AWS_REGION` must be set.

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [x] Release notes are filled out
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: test/integration/auth/mongodb_aws.test.ts [hunk 1]
```diff
@@ -88,4 +88,154 @@ describe('MONGODB-AWS', function () {
...
+            process.env.AWS_REGION = test.env.AWS_REGION;
+          }
+
+          calledArguments = [];
+          MongoDBAWS.credentialProvider = {
+            fromNodeProviderChain(...args) {
+              calledArguments = args;
+              return credentialProvider.fromNodeProviderChain(...args);
+            }
+          };
```
Comment (nbbeeken) [inline test/integration/auth/mongodb_aws.test.ts:205]: Ideally, we can find another mechanism to monitor the actual destination of the http request to complete the "integration" test.

**Comment by nbbeeken** [comment]
Looking at the tests on this task:

[aws-6.0-auth-test-run-aws-auth-test-AssumeRoleWithWebIdentity-with-AWS_ROLE_SESSION_NAME-set](https://spruce.mongodb.com/task/mongo_node_driver_next_ubuntu1804_test_mongodb_aws_aws_6.0_auth_test_run_aws_auth_test_AssumeRoleWithWebIdentity_with_AWS_ROLE_SESSION_NAME_set_patch_a7ffdf5aca888d87dc5c8d9661891a6ca3121c36_64e3ac8fc9ec44e563bb571c_23_08_21_18_27_27?execution=0&sortBy=STATUS&sortDir=ASC)

File: src/cmap/auth/mongodb_aws.ts [hunk 2]
```diff
@@ -34,6 +57,7 @@ interface AWSSaslContinuePayload {
 }
 
 export class MongoDBAWS extends AuthProvider {
+  static credentialProvider: ReturnType<typeof getAwsCredentialProvider> | null = null;
```
Comment (nbbeeken) [inline src/cmap/auth/mongodb_aws.ts:60]: This serves two purposes, 1. we usually cache our lazy imports so we don't call require more than once, 2. I can now access the SDK this auth code uses for testing purposes

Packages (this includes the driver!) that compile with typescript have all the properties on their `module.exports` defined as non-configurable getters. While this emulates ES Module live binding and immutability, it makes sinon spying impossible.

**Review by W-A-James — COMMENTED** [review]
Apart from the one question on the test file (which I won't block on), LGTM

File: test/integration/auth/mongodb_aws.test.ts [hunk 3]
```diff
@@ -88,4 +88,154 @@ describe('MONGODB-AWS', function () {
...
+          const { AWS_WEB_IDENTITY_TOKEN_FILE = '' } = process.env;
+          credentialProvider = (() => {
+            try {
+              return require('@aws-sdk/credential-providers');
+            } catch {
+              return null;
+            }
+          })();
+          return AWS_WEB_IDENTITY_TOKEN_FILE.length === 0 || credentialProvider == null;
+        };
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/integration/auth/mongodb_aws.test.ts:177]: It can be, I'll fix this

File: test/integration/auth/mongodb_aws.test.ts [hunk 4]
```diff
@@ -88,4 +88,154 @@ describe('MONGODB-AWS', function () {
...
+        title: 'uses the region from the environment',
+        env: {
+          AWS_STS_REGIONAL_ENDPOINTS: 'legacy',
+          AWS_REGION: 'us-west-2'
+        },
+        outcome: null
+      },
+      {
+        ctx: 'when AWS_STS_REGIONAL_ENDPOINTS is set to legacy and region is new',
+        title: 'uses the region from the environment',
```
Comment (durran) [inline test/integration/auth/mongodb_aws.test.ts:152]: Should this be "uses the default region" ?

**Review by durran — CHANGES_REQUESTED** [review]

**Review by W-A-James — APPROVED** [review]

**Pull request merged**